app/build_svc.py

```diff
@@ -19,78 +20,180 @@ def __init__(self, services):
+    async def _stage_docker_directory(self, ability):
+        """Create code file and DLL references in Docker build directory
+
+        :param ability: Ability to be built
+        :type ability: Ability
+        """
         if ability.code:
             with open(os.path.join(self.build_directory, ability.language, self.build_file), 'w') as f:
                 f.write(self.decode_bytes(ability.code))
 
-    def _purge_build_folder(self, language):
+        for payload in [p for p in ability.payloads if p.endswith('.dll')]:
+            payload_name = payload
+            if self.is_uuid4(payload):
+                payload_name, _ = self.file_svc.get_payload_name_from_uuid(payload)
+            _, src = await self.file_svc.find_file_path(payload_name)
+            dst = os.path.join(self.build_directory, ability.language)
+            shutil.copy(src=src, dst=dst)
+
+    def _purge_build_directory(self, language):
+        """Remove all files from language build directory
+
+        :param language: Language to clear build directory for
+        :type language: string
+        """
         for f in glob.iglob(f'{self.build_directory}/{language}/*'):
             os.remove(f)
 
     def _run_target_docker(self, ability, args):
+        """Run docker container to build ability files
+
+        :param ability: Ability to build code for
+        :type ability: Ability
+        :param args: Arguments to pass to the entrypoint command
+        :type args: string
+        """
         env = self.get_config(prop='enabled', name='build').get(ability.language)
         container = self.docker_client.containers.run(image=self.build_envs[ability.language].short_id, remove=True,
-                                                      command=('%s %s' % (env['entrypoint'], args)).split(' '),
+                                                      command='{} {}'.format(env['entrypoint'], args).split(' '),
                                                       working_dir=env['workdir'],
+                                                      user=os.getuid(),
                                                       volumes={os.path.abspath(
                                                                os.path.join(self.build_directory, ability.language)):
                                                                dict(bind=env['workdir'], mode='rw')}, detach=True)
         code = container.wait()
-        self.log.debug('Container for %s ran for ability ID %s: %s' % (ability.language, ability.ability_id,
-                                                                       code))
+        self.log.debug('Container for {} ran for ability ID {}: {}'.format(ability.language, ability.ability_id, code))
+
+    def _check_errors(self, language):
+        """Check for errors which occurred during the build
+
+        :param language: Language to check errors for
+        :type language: string
+        """
+        if language == 'csharp':
```

Comment:
Have you thought about how this function would work if builder started using other languages? What if we introduced C++ exploits?
Explicitly calling out 'csharp' might work now, but looks harder to scale.

Reply:
We wouldn't use the `csc` compiler for other languages, so the output format would be different. If we wanted error checking in another compiler, we would need to write a parser for that language / compiler. I'm not sure how we could "scale" this.